### ISSUE TITLE
PatternMatchingInstanceof leaves unnecessary parentheses

### ISSUE THREAD

**Issue by PhilippWendler** [google/error-prone#4996]
Since ab45284 the `PatternMatchingInstanceof` check leaves parentheses that are flagged by the UnnecessaryParentheses check. I guess it is not desired that one Error Prone check produces code that another check than complains about?

One particular important case is:
```java
if (o instanceof Test) {
  Test t = ((Test) o);
}
```
The parentheses make Error Prone miss the fact that it can eliminate the variable declaration and directly use `o instanceof Test t`.

Other cases found inside our code case and flagged by UnnecessaryParentheses are in return statements, inside divisions, and ternary operators. It is easy to extend `PatternMatchingInstanceof.requiresParentheses()` and fix particular cases, and I could do so in a PR, but I wonder whether this is the best approach, or whether the function should instead use a list of cases with required parentheses? So I wanted to clarify this upfront.

**Comment by cushon** [comment]
It would be nice if PatternMatchingInstanceof handled all these cases. Would you not get the same result by running UnnecessaryParentheses first? Maybe that check does things you don't want?

> I wonder whether this is the best approach, or whether the function should instead use a list of cases with required parentheses?

Maybe, I'm not sure. The risk of checking for cases where parentheses are required is that if we miss them the check produces bad findings, instead of missing a finding. Either way I find it hard to remember all of the cases where they are/aren't required. I'm open to being convinced that handling the cases where they're required is simpler :)

**Comment by graememorgan** [comment]
> PatternMatchingInstanceof.requiresParentheses

There's no such method, so I'm not quite sure what you mean. I mailed a fix though.

I'd be in favour of UnnecessaryParentheses firing on `foo = (a);` even for the case in the above comment, but I can see that being controversial.

**Comment by PhilippWendler** [comment]
> > PatternMatchingInstanceof.requiresParentheses
> 
> There's no such method, so I'm not quite sure what you mean. I mailed a fix though.

I was referring to this:

https://github.com/google/error-prone/blob/b20b0195ec27b5886cacaf9f37e19512280e634e/core/src/main/java/com/google/errorprone/bugpatterns/PatternMatchingInstanceof.java#L272

If you have a fix, great, I am willing to test.

**Comment by graememorgan** [comment]
Apparently I should have gone to Specsavers. :)
